### PR TITLE
Adds swerve drive to drivetrains

### PR DESCRIPTION
**Pull request by CarterFendley**
This is the version of swerve physics that works with team 1418's 2017 robot. 

This _should_ be universal however there is a possibility that I made two mistakes in my math that cancel each other out.

### CODE REVIEW

File: lib/pyfrc/physics/drivetrains.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ def four_motor_swerve_drivetrain(lr_motor, rr_motor, lf_motor, rf_motor, lr_angl
     rf_rad = rf_angle * (math.pi / 180)
 
     # Calculate wheelbase radius
-    wheelbase_radius = math.sqrt((x_wheelbase ** 2) + (y_wheelbase ** 2))
+    wheelbase_radius = math.sqrt(((x_wheelbase/2) ** 2) + ((y_wheelbase/2) ** 2))
```
Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: You can get rid of two pairs of parentheses here. Order of operations should be obvious to anyone who's done… well, any maths.

(Seeing more than two consecutive open parentheses weirds me out. Maybe I'm just allergic to Lisp.)

Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: Yea I probably should have less parentheses. My default is always just through them in because they never hurt.

Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: I'd personally argue that having superfluous parentheses when order of operations is obvious, such as in this scenario, hurts readability. I had to keep glancing left and right to work out what the order of operations is, when it's just normal maths here.

Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: Oh yea, it defiantly hurts readability. But I have worked with some APIs before that just suck at order of operations so it became a habit

File: lib/pyfrc/physics/drivetrains.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import math
```
Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: This import should go _under_ the docstring — having statements before a docstring makes the docstring… not a docstring.

File: lib/pyfrc/physics/drivetrains.py [hunk 3]
```diff
@@ -140,6 +140,64 @@ def mecanum_drivetrain(lr_motor, rr_motor, lf_motor, rf_motor, x_wheelbase=2, y_
...
+    rf = rf_motor * speed
+
+    # Calculate angle in radians
+    lr_rad = lr_angle * (math.pi / 180)
+    rr_rad = rr_angle * (math.pi / 180)
+    lf_rad = lf_angle * (math.pi / 180)
+    rf_rad = rf_angle * (math.pi / 180)
+
+    # Calculate wheelbase radius
+    wheelbase_radius = math.sqrt((x_wheelbase / 2) ** 2 + (y_wheelbase/2) ** 2)
```
Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: Was unsure weather it would do `x_wheelbase / 2^2` so I left in those parentheses

Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: Well, if you removed those parens, yes, it would definitely not do what you wanted. I would assume you remember order of operations from school (I believe the common mnemonic in the US is PEMDAS?). 😜

Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: Welll this is the reason I use parentheses... So I don't have to spend 30 secs looking through my long math statement to make sure it follows order of operations.

I just through parentheses in there 😜

File: lib/pyfrc/physics/drivetrains.py [hunk 4]
```diff
@@ -140,6 +140,64 @@ def mecanum_drivetrain(lr_motor, rr_motor, lf_motor, rf_motor, x_wheelbase=2, y_
...
+    lr = lr_motor * speed
+    rr = rr_motor * speed
+    lf = lf_motor * speed
+    rf = rf_motor * speed
+
+    # Calculate angle in radians
+    lr_rad = lr_angle * (math.pi / 180)
+    rr_rad = rr_angle * (math.pi / 180)
+    lf_rad = lf_angle * (math.pi / 180)
+    rf_rad = rf_angle * (math.pi / 180)
```
Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: Probably best to either:
* have the user specify the angles in radians (not sure whether this would work though), or
* use `math.radians` instead of manually multiplying by π/180.

Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: So `math.radians(x)` is nice but is there any reason it would be _better_ than manually doing it?

Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: Not every high school student knows that π/180 is the degrees-to-radians conversion ratio.

Anyway, having read the code a bit more, I realised that angles are returned in radians. It'll be quite confusing to give the function angles in degrees and get radians out. Probably best to keep everything to radians.

Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: Where is it returning radians?

File: lib/pyfrc/physics/drivetrains.py (hunk 3 above)
Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: Well, if your statements are so long that you're spending 30 seconds staring at it… maybe you should be breaking it up into multiple (logical) statements.

Comment (CarterFendley) [inline lib/pyfrc/physics/drivetrains.py]: Yea this was sorta a port from my robot code this year... Since I am the only programmer there every day sometimes things get sloppy because I have a lot to do.

Comment (virtuald) [inline lib/pyfrc/physics/drivetrains.py]: I prefer having parentheses because I don't actually remember order of operations without looking it up.

File: lib/pyfrc/physics/drivetrains.py (hunk 4 above)
Comment (virtuald) [inline lib/pyfrc/physics/drivetrains.py]: I prefer using math.radians, as it's more explicit.

However, I agree with @auscompgeek , it is very strange that your input is in degrees and your output is in radians. Best to do both in radians unless there's a compelling reason not to do so.

File: lib/pyfrc/physics/drivetrains.py (hunk 3 above)
Comment (virtuald) [inline lib/pyfrc/physics/drivetrains.py]: You're just calculating the hypotenuse? Use `math.hypot` instead, it handles various edge cases correctly.

**Comment by virtuald** [comment]
This is really useful stuff, thanks for submitting your PR! Once the issues above are addressed, then we can merge it.

Could you add some very simple tests to check that your output is correct? I know, I know... but you'll be happier in the end if they catch a bug you hadn't thought of.

I added tests for the existing drivetrains at https://github.com/robotpy/pyfrc/blob/master/tests/test_drivetrains.py

File: lib/pyfrc/physics/drivetrains.py (hunk 2 above)
Comment (auscompgeek) [inline lib/pyfrc/physics/drivetrains.py]: Hmm. GitHub seems to want to keep showing this outdated diff.

Anyway, good thing you've fixed the issue here.

**Comment by virtuald** [comment]
@CarterFendley would love to see this finished

**Comment by CarterFendley** [comment]
I have fixed the issues with this PR. Having some issues pushing to my repo. Will make a new PR soon.

**Comment by virtuald** [comment]
You can just force push to your branch, instead of opening a new PR.

**Comment by CarterFendley** [comment]
Yea it tried that. I was having an issue where it told me that everything
was up to date.

**Comment by virtuald** [comment]
Then you didn't commit your changes.

**Comment by CarterFendley** [comment]
That's the thing. I am pretty sure I did. The logs said I did.

**Comment by virtuald** [comment]
Did you push the correct branch?

**Comment by CarterFendley** [comment]
Okay I sorted it out. Still don't know what the original issue was.

**Comment by virtuald** [comment]
Sorry for not reviewing this yet, I've been busy moving. Maybe later next week.

**Comment by virtuald** [comment]
This is great work, thanks!
